### DISCUSSION TITLE
Problem seen when installing an BPF program

### DISCUSSION THREAD

**Discussion by josyulav** [bpfman/bpfman#818] [Q&A]
I do not know where to post, I am an newbee....

Setup
1 Standalone VM
2.Os running : 
[root@localhost release]# uname -mrs
Linux 4.18.0-477.27.1.rt7.290.el8_8.x86_64+debug x86_64
[root@localhost release]#

Steps : 
1) start release version bpfd
[root@localhost release]# pwd
/home/vjosyula/src/bpfd-0.3.0/target/x86_64-unknown-linux-musl/release
[root@localhost release]# RUST_LOG=debug bpfd
2) installing program using bpfctl
[root@localhost bpfd]# pwd
/home/vjosyula/go/src/github.com/bpfd-dev/bpfd
[root@localhost bpfd]# ./target/debug/bpfctl load-from-image --image-url quay.io/bpfd-bytecode/xdp_pass:latest xdp --iface lo --priority 50

Error seen on bpfd side
[DEBUG aya_obj::relocation] finished relocating program `xdp_dispatcher` function `compat_test`
[DEBUG aya_obj::relocation] callee `compat_test` is at ins 208, 62 from current instruction 145
[DEBUG aya_obj::relocation] finished relocating program `xdp_dispatcher` function `xdp_dispatcher`
[DEBUG aya_obj::relocation] relocating program `LBB11_22` function `LBB11_22` size 0
[DEBUG aya_obj::relocation] finished relocating program `LBB11_22` function `LBB11_22`
[DEBUG bpfd::multiprog::xdp] XdpDispatcher::attach_extensions() for if_index 1, revision 1
[DEBUG aya_obj::btf::btf] [DATASEC] .rodata: fixup size to 51
[DEBUG aya_obj::btf::btf] [DATASEC] .rodata: VAR GLOBAL_u8: fixup offset 0
[DEBUG aya_obj::btf::btf] [DATASEC] .rodata: VAR GLOBAL_u32: fixup offset 4
[DEBUG aya_obj::btf::btf] [DATASEC] .rodata: VAR pass.____fmt: fixup not required
[DEBUG aya_obj::btf::btf] [DATASEC] license: fixup size to 13
[DEBUG aya_obj::btf::btf] [DATASEC] license: VAR _license: fixup offset 0
[DEBUG aya_obj::relocation] relocating map by section index 5, kind Rodata at insn 3 in section 3
[DEBUG aya_obj::relocation] relocating map by section index 5, kind Rodata at insn 6 in section 3
[DEBUG aya_obj::relocation] relocating map by section index 5, kind Rodata at insn 0 in section 3
[DEBUG aya_obj::relocation] relocating program `pass` function `pass` size 12
[DEBUG aya_obj::relocation] finished relocating program `pass` function `pass`
[DEBUG bpfd::multiprog::xdp] XdpDispatcher::attach() for if_index 1, revision 1
thread 'main' panicked at bpfd/src/multiprog/xdp.rs:164:18:
called `Result::unwrap()` on an `Err` value: InvalidLink
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[WARN  bpfd::rpc] RPC load error: RecvError(
        (),
    )
[DEBUG trust_dns_proto::xfer::dns_exchange] io_stream is done, shutting down
[root@localhost release]# 

Output on terminal where bpfd is run
[root@localhost bpfd]# ./target/debug/bpfctl load-from-image --image-url quay.io/bpfd-bytecode/xdp_pass:latest xdp --iface lo --priority 50
Error = status: Unknown, message: "transport error", details: [], metadata: MetadataMap { headers: {} }

Caused by:
    0: transport error
    1: connection error: unexpected end of file
    2: unexpected end of file
Error: Failed to execute request
[root@localhost bpfd]#

Anyhelp in the right direction will be useful

**Comment by Billy99** [comment]
Hey @josyulav,
Thanks for your interest in bpfd. As far as where to post, here is fine, or create a new issue. You could also post on our slack on Kubernetes.slack.com and #bpfd (https://kubernetes.slack.com/archives/C04UJBW2553).

I went to play look at your issues yesterday and there were some Quay.io issues that side tracked me. My first thought is that you need a newer kernel. We haven't posted a minimum required kernel yet, but eBPF in general depends on kernel features. One thing may be broken on an older kernel but something else may work, so sometimes it depends on what you are doing. But 4.18 is pretty old from an eBPF perspective. If you are just trying to learn eBPF, it will make your life easier to upgrade to a newer kernel. If you require that kernel, need to determine what features you need and then what kernel version supports those feature.

**Comment by josyulav** [comment]
@Billy99 ,

I saw the same issues when trying to run unit tests as well. So are you suggesting that kernel version on which I running is pretty old for it to work ?

I am not trying learn eBPF, we want something to manage eBPF programs. So we are exploring bpfd....

Also, I do not have an access to the slack channel, I do not have an account there. To create an account, it is asking me to get invitation. Not sure how to get an inivitation for the same.

Thanks,
Regards,


Compiling form_urlencoded v1.2.0
   Compiling assert_matches v1.5.0
   Compiling difflib v0.4.0
   Compiling termtree v0.4.1
   Compiling bitflags v2.4.0
   Compiling url v2.4.1
   Compiling predicates v3.0.4
   Compiling aya v0.11.0 (https://github.com/aya-rs/aya?branch=main#62849944)
   Compiling tonic v0.10.2
   Compiling predicates-tree v1.0.9
   Compiling clap v4.4.8
   Compiling toml v0.7.8
   Compiling bstr v1.7.0
   Compiling wait-timeout v0.2.0
   Compiling regex v1.10.0
   Compiling assert_cmd v2.0.12
   Compiling integration-test-macros v0.1.0 (/home/vjosyula/go/src/[github.com/bpfd-dev/bpfd/tests/integration-test-macros](http://github.com/bpfd-dev/bpfd/tests/integration-test-macros))
   Compiling env_logger v0.10.1
   Compiling inventory v0.3.13
   Compiling bpfd-api v0.3.0 (/home/vjosyula/go/src/[github.com/bpfd-dev/bpfd/bpfd-api](http://github.com/bpfd-dev/bpfd/bpfd-api))
   Compiling integration-test v0.1.0 (/home/vjosyula/go/src/[github.com/bpfd-dev/bpfd/tests/integration-test](http://github.com/bpfd-dev/bpfd/tests/integration-test))
    Finished dev [unoptimized + debuginfo] target(s) in 28.91s
[INFO  integration_test] Executing test case(s): ["test_load_unload_xdp", "test_proceed_on_xdp"]
[INFO  integration_test] Running integration_test::tests::basic::test_load_unload_xdp
[DEBUG integration_test::tests::utils] Successfully created namespace bpfd-int-test
[DEBUG integration_test::tests::utils] sucessfully started ping to namespace bpfd-int-test at address 172.37.37.2
[DEBUG integration_test::tests::utils] Starting bpfd
[DEBUG integration_test::tests::utils] Successfully Started bpfd
[DEBUG integration_test::tests::basic] Installing xdp_pass programs
thread 'main' panicked at bpfd/src/multiprog/xdp.rs:164:18:
called `Result::unwrap()` on an `Err` value: InvalidLink
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at tests/integration-test/src/tests/utils.rs:182:43:
called `Result::unwrap()` on an `Err` value: OutputError { cmd: Some("\"/home/vjosyula/go/src/[github.com/bpfd-dev/bpfd/target/debug/bpfctl\](http://github.com/bpfd-dev/bpfd/target/debug/bpfctl%5C)" \"load-from-image\" \"--global\" \"GLOBAL_u8=61\" \"GLOBAL_u32=0D0C0B0A\" \"--image-url\" \"[quay.io/bpfd-bytecode/xdp_pass:latest\](http://quay.io/bpfd-bytecode/xdp_pass:latest%5C)" \"--pull-policy\" \"Always\" \"xdp\" \"--iface\" \"veth-bpfd-host\" \"--priority\" \"111\" \"--proceed-on\" \"aborted\" \"drop\" \"pass\" \"tx\" \"redirect\" \"dispatcher_return\""), stdin: None, cause: Expected(Output { output: Output { status: ExitStatus(unix_wait_status(256)), stdout: "", stderr: "Error = status: Aborted, message: \"channel closed\", details: [], metadata: MetadataMap { headers: {\"content-type\": \"application/grpc\", \"date\": \"Tue, 14 Nov 2023 18:20:16 GMT\", \"content-length\": \"0\"} }\nError: Failed to execute request\n" } }) }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[DEBUG integration_test::tests::utils] stopping bpfd
[DEBUG integration_test::tests::utils] stopping ping
[DEBUG integration_test::tests::utils] namespace bpfd-int-test deleted
[root@localhost bpfd]#

Can you point me what could be wrong ?

Steps 

1. Build bpfd by clone project. Then ran below command,
2. RUST_LOG=debug cargo xtask integration-test -- test_load_unload_xdp test_proceed_on_xdp

**Comment by Billy99** [comment]
@josyulav
Sorry for the delay. At one point, there was work being done in AYA (which is used by bpfd) to detect the features supported by the kernel and report it back to allow us to determine dynamically what features bpfd could leverage. With backports in different distros, just using a hardcoded kernel number can be problematic. That feature is still on the roadmap, but fell lower in priority for us.

That being said, it is probably best to us a 5.9 kernel or greater. Two of the main kernel features leveraged by bpfd is the extension support, which is what allows bpfd to load multiple XDP and TC programs on a single interface. And bpf_link for perf_attach programs (Kprobe/Uprobe/Tracepoint etc...). To use an older kernel would require looking through older kernels for backports, and testing to see if there are other features that we might rely on. Longer term, we would like to use the feature mentioned above to be more robust on the kernels and work with limited features on older kernels. But that is not there today.

The 5.9 kernel is based on looking through code and our best guess. I would need to create a 5.9 based VM and verify there aren't any other surprises. Before I do that, are you needing the oldest kernel we support, or is it ok to use a newer kernel for initial investigation and to kick the tires?

Last, we have our community meeting every Thursday at 10:00 AM Eastern Time. Feel free to join and ask us questions.(https://meet.google.com/ggz-zkmp-pxx)

**Comment by josyulav** [comment]
Hi Billy,

Thanks for your reply. My comments are inline
>I would need to create a 5.9 based VM and verify there aren't any other
surprises. Before I do that, are you needing the oldest kernel we support,
or is it ok to use a newer kernel for initial investigation and to kick the
tires?
I have rocky 9.2 based VM, I verified it and it works and it is fine.
Question
I too looked at the code. I also similar suspicion about the same. Only
thing I could not understand is why are trying to use bpf_link to pin it ?
That part I did not understand. Can you educate me on this too ?

**Comment by josyulav** [comment]
I forgot to answer one of your questions, It would nice, I can know the
oldest kernel supported.

But looking at the bpf_link feature, that came to kernel 5.6 - 5.7 kernel.
I looked link
https://elixir.bootlin.com/linux/v5.5.19/source/tools/lib/bpf/libbpf.c#L5471

So only if we upgrade to 5.7 kernel we will fully get the support.

**Comment by Billy99** [comment]
@josyulav We use the pinning so the eBPF program stays installed if the loading process (bpfd) goes away or is restarted.

**Comment by josyulav** [comment]
Hi Billy,

Another question when using below
bpfctl load-from-file --path ...

Questions
1) the file passed as parameter should it always use bpf2go to compile your
c code ? Can clang not be used
2) can eBPF code compiled in rust can be used ?

Thanks,
Regards,
Venu

**Comment by dave-tucker** [comment]
> Questions
> 1) the file passed as parameter should it always use bpf2go to compile your
c code ? Can clang not be used

Both bpf2go AND clang can be used to compile your bytecode

> 2) can eBPF code compiled in rust can be used ?

Rust won't work for XDP/TC programs yet since bpf-linker doesn't output BTF for Rust Programs although it's being worked on.
It will work for uprobes/kprobes/tracepoints.
